### PR TITLE
Jax 2d

### PR DESCRIPTION
**Pull request by MuawizChaudhary**
This PR includes the jax implementation of the 2d scattering transform.

### CODE REVIEW

File: kymatio/backend/jax_backend.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        raise TypeError('The first input must be complex.')
+
+    if A.shape[-len(B.shape):] != B.shape[:]:
+        raise RuntimeError('The inputs are not compatible for '
+                           'multiplication.')
+
+    if not _is_complex(B) and not _is_real(B):
+        raise TypeError('The second input must be complex or real.')
+
+    if inplace:
```
Comment (eickenberg) [inline kymatio/backend/jax_backend.py:77]: maybe don't add inplace back if we are trying to get rid of it in the package

Comment (MuawizChaudhary) [reply kymatio/backend/jax_backend.py:77]: Yeah, I can remove it now or after the inplace pr is merged.

File: kymatio/scattering2d/backend/jax_backend.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
+# Authors: Edouard Oyallon, Sergey Zagoruyko, Muawiz Chaudhary
```
Comment (eickenberg) [inline kymatio/scattering2d/backend/jax_backend.py:1]: this author list is wrong. we should probably remove all those lines from the full repo, because they do not give any information beyond the authors and contributors list we already have. so i'd remove this here and then do a separate PR to remove it everywhere else

File: tests/scattering2d/test_jax_scattering2d.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+backends.append(backend)
+
+
+class TestPad:
+    @pytest.mark.parametrize('backend', backends)
+    def test_Pad(self, backend):
+        pad = backend.Pad((2, 2, 2, 2), (4, 4), pre_pad=False)
+
+        x = np.random.randn(4, 4) + 1J * np.random.randn(4, 4)
+        x = x[np.newaxis, ...]
```
Comment (eickenberg) [inline tests/scattering2d/test_jax_scattering2d.py:22]: why not just add the axis above when creating `x`

File: kymatio/scattering2d/backend/jax_backend.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    def __init__(self, pad_size, input_size, pre_pad=False):
+        """
+            Padding which allows to simultaneously pad in a reflection fashion
+            and map to complex.
+
+            Parameters
+            ----------
+            pad_size : list of 4 integers
+                size of padding to apply.
+            input_size : list of 2 integers
```
Comment (eickenberg) [inline kymatio/scattering2d/backend/jax_backend.py:23]: what do we need this arg for? It seems to get completely lost. Do other backends need it?

File: tests/scattering2d/test_jax_scattering2d.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+        assert z[0, 1, 2] == x[0, 1, 0]
+        assert z[0, 1, 3] == x[0, 1, 1]
+
+        pad = backend.Pad((2, 2, 2, 2), (4, 4), pre_pad=True)
+
+        x = np.random.randn(8, 8) + 1J * np.random.randn(8, 8)
+
+        z = pad(x)
+
+        assert np.allclose(x, z)
```
Comment (eickenberg) [inline tests/scattering2d/test_jax_scattering2d.py:39]: this doesn't seem like it can pass

File: tests/scattering2d/test_jax_scattering2d.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+
+class TestModulus:
+    @pytest.mark.parametrize('backend', backends)
+    def test_Modulus(self, backend):
+        modulus = backend.modulus
+
+        x = np.random.rand(100, 10, 4) + 1J * np.random.rand(100, 10, 4)
+
+        y = modulus(x)
+        u = np.squeeze(np.sqrt(np.real(x) ** 2 + np.imag(x) ** 2))
```
Comment (eickenberg) [inline tests/scattering2d/test_jax_scattering2d.py:60]: what is the squeeze for?

I would opt for `x.real, x.imag`

File: tests/scattering2d/test_jax_scattering2d.py [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+        v = y
```
Comment (eickenberg) [inline tests/scattering2d/test_jax_scattering2d.py:61]: that seems unnecessary

File: kymatio/scattering2d/backend/jax_backend.py (hunk 2 above)
Comment (MuawizChaudhary) [reply kymatio/scattering2d/backend/jax_backend.py:1]: agree

File: kymatio/scattering2d/backend/jax_backend.py (hunk 4 above)
Comment (MuawizChaudhary) [reply kymatio/scattering2d/backend/jax_backend.py:23]: I believe torch needs it. https://github.com/kymatio/kymatio/blob/dev/kymatio/scattering2d/backend/torch_backend.py#L76

File: tests/scattering2d/test_jax_scattering2d.py (hunk 3 above)
Comment (MuawizChaudhary) [reply tests/scattering2d/test_jax_scattering2d.py:22]: Good question.

File: tests/scattering2d/test_jax_scattering2d.py (hunk 5 above)
Comment (MuawizChaudhary) [reply tests/scattering2d/test_jax_scattering2d.py:39]: it passes

File: tests/scattering2d/test_jax_scattering2d.py (hunk 6 above)
Comment (MuawizChaudhary) [reply tests/scattering2d/test_jax_scattering2d.py:60]: the squeeze is unnecessary

File: kymatio/scattering2d/backend/jax_backend.py (hunk 4 above)
Comment (eickenberg) [reply kymatio/scattering2d/backend/jax_backend.py:23]: I don't actually understand why it can't get that out of the batch. it's not like it's precomputing anything

File: tests/scattering2d/test_jax_scattering2d.py (hunk 5 above)
Comment (eickenberg) [reply tests/scattering2d/test_jax_scattering2d.py:38]: how? `z=pad(x)` can't even be of the same shape as `x`

Comment (eickenberg) [reply tests/scattering2d/test_jax_scattering2d.py:38]: ah, it's `pre_pad=False` -- does this mean it's already padded or not? Not, right? So then they should be different shapes. I guess we wanted to get rid of that `pre_pad` thing anyway
